### ISSUE TITLE
TransferUtility upload - Support InputStream and FileDescriptor instead of only File

### ISSUE THREAD

**Issue by jlamson** [aws-amplify/aws-sdk-android#1436]
**Which AWS Services is the feature request for?**
S3, specifically the upload functionality provided by `TransferUtility`

**Is your feature request related to a problem? Please describe.**
When allowing users to pick files on Android, it is incredible rare you are given an instance of a `File` itself. Using the standard [Storage Access Framework (SAF)](https://developer.android.com/guide/topics/providers/document-provider) to allow the user to pick a file from any source on their device, you're returned a `content://` URI. This URI is not reliably a file. Via a `ContentProvider`, you can however get access to an InputStream or a [DocumentsProvider](https://developer.android.com/reference/android/provider/DocumentsProvider.html) that can provide `FileDescriptor` objects. 

[This article](https://commonsware.com/blog/2016/03/15/how-consume-content-uri.html) from CommonsWare has the most succinct explanation i've found on how to read from Content URIs. None of the solutions provided are compatible with the `TransferUtility::upload` functions. Instead, you must copy the entire `File` from the input stream, then pass that to the `TransferUtility`. This is inconvenient and waste the user's time/resources. 

**Describe the solution you'd like**
In short, `TransferUtility` should allow the upload of content directly from an `InputStream` or a `FileDescriptor`. With the way Android's behaviour of inter-app content is heading, `File` is no longer a good object on which to base your API. 

**Describe alternatives you've considered**
Mentioned above, the workaround here is to either interface with the S3 REST APIs directly, or copy the user's content to the app's directories, then upload that.

**Comment by marcin-adamczewski** [comment]
Guys, you should definitely prioritize it. As @jlamson mentioned you now have to copy the whole file in order to upload it when you take it from the system file picker. Imagine it is a big file and user have no disk space for such copying. It is also cumbersome to use.
As you have to store the handle to the file in database for resuming operations, etc. maybe it would be better if  `TransferUtility::upload` method take `Uri` as a param instead of inputstream? That way you can store a handle in database for further use. And using content resolver you can get all the info you need from it. I suppose it may be also required that the provided Uri has `persistableUriPermission` so you can resume uploads after re-booting device.

**Comment by changxu0306** [comment]
Hello! Now we support uploading `InputStream` using `TransferUtility` in [AWS Android SDK v2.19.2](https://github.com/aws-amplify/aws-sdk-android/releases/tag/release_v2.19.2). You can just call `upload()` and pass your `InputStream` as a parameter to the method. The corresponding API in Amplify framework will be released in the next version. I'll close this issue for now. Feel free to reopen it if you have further problems. Thank you for the suggestion!
